src/server/frontend/wayland/wayland_connector.cpp

```diff
@@ -1954,6 +1961,532 @@ class WlShell : public wayland::Shell
+    }
+
+    std::shared_ptr<mf::Shell> const shell;
+    XdgSurfaceV6* const self;
+};
+
+struct XdgPopupV6 : wayland::XdgPopupV6
+{
+    XdgPopupV6(struct wl_client* client, struct wl_resource* parent, uint32_t id) :
+        wayland::XdgPopupV6(client, parent, id)
+    {
+    }
+
+    void grab(struct wl_resource* seat, uint32_t serial) override
+    {
+        (void)seat, (void)serial;
+        // TODO
+    }
+
+    void destroy() override
+    {
+        wl_resource_destroy(resource);
+    }
+};
+
+class XdgSurfaceV6EventSink : public BasicSurfaceEventSink
+{
+public:
+    using BasicSurfaceEventSink::BasicSurfaceEventSink;
+
+    XdgSurfaceV6EventSink(WlSeat* seat, wl_client* client, wl_resource* target, wl_resource* event_sink,
+        std::shared_ptr<bool> const& destroyed) :
+        BasicSurfaceEventSink(seat, client, target, event_sink),
+        destroyed{destroyed}
+    {
+        auto const serial = wl_display_next_serial(wl_client_get_display(client));
+        post_configure(serial);
+    }
+
+    void send_resize(geometry::Size const& new_size) const override
+    {
+        if (window_size != new_size)
+        {
+            auto const serial = wl_display_next_serial(wl_client_get_display(client));
+            notify_resize(new_size);
+            post_configure(serial);
+        }
+    }
+
+    std::function<void(geometry::Size const& new_size)> notify_resize = [](auto){};
+
+private:
+    void post_configure(int serial) const
+    {
+        seat->spawn(run_unless(destroyed, [event_sink= event_sink, serial]()
+            {
+                wl_resource_post_event(event_sink, 0, serial);
+            }));
+    }
+
+    std::shared_ptr<bool> const destroyed;
+};
+
+struct XdgSurfaceV6 : wayland::XdgSurfaceV6, WlAbstractMirWindow
+{
+    XdgSurfaceV6* get_xdgsurface(wl_resource* surface) const
+    {
+        auto* tmp = wl_resource_get_user_data(surface);
+        return static_cast<XdgSurfaceV6*>(static_cast<wayland::XdgSurfaceV6*>(tmp));
+    }
+
+    XdgSurfaceV6(wl_client* client,
+        wl_resource* parent,
+        uint32_t id,
+        wl_resource* surface,
+        std::shared_ptr<mf::Shell> const& shell,
+        WlSeat& seat) :
+        wayland::XdgSurfaceV6(client, parent, id),
+        WlAbstractMirWindow{client, surface, resource, shell},
+        parent{parent},
+        shell{shell},
+        sink{std::make_shared<XdgSurfaceV6EventSink>(&seat, client, surface, resource, destroyed)}
+    {
+        WlAbstractMirWindow::sink = sink;
+    }
+
+    ~XdgSurfaceV6() override
+    {
+        auto* const mir_surface = get_wlsurface(surface);
+        mir_surface->set_role(&nullwlmirwindow);
+    }
+
+    void destroy() override
+    {
+        wl_resource_destroy(resource);
+    }
+
+    void get_toplevel(uint32_t id) override
+    {
+        new XdgToplevelV6{client, parent, id, shell, this};
+        auto* const mir_surface = get_wlsurface(surface);
+        mir_surface->set_role(this);
```

Comment:
I'm finding this hard to grok. Method doesn't appear to "get" anything. is it a bad name due to the autogenerated interface that's out of our control?

Anyway this creates a XdgToplevelV6 on the heap that associates itself to the XdgSurfaceV6 as a user_data resource, adding a resize handler and the wayland resource implementation. Yes?

Reply:
Yes. It's all described in /usr/share/wayland-protocols/unstable/xdg-shell/xdg-shell-unstable-v6.xml

I don't like this naming, but I think this is the least confusing approach.